### DISCUSSION TITLE
On Oracle, with Identity() primary key, ORM fails with "Instance has a NULL identity key"

### DISCUSSION THREAD

**Discussion by koenigdavidmj** [sqlalchemy/sqlalchemy#6909] [Usage Questions] [answered]
### Describe the bug

I'm trying to run against Oracle 12.1.0.2, which I believe is new enough that Identity() primary key columns work. However, it fails with:

> sqlalchemy.orm.exc.FlushError: Instance <Bleh at (addr)> has a NULL identity key.  If this is an auto-generated value, check that the database table allows generation of new primary key values, and that the mapped Column object is configured to expect these generated values.  Ensure also that this flush() is not occurring at an inappropriate time, such as within a load() event.

### To Reproduce

```python
#!/usr/bin/env python
import os
 
import cx_Oracle
cx_Oracle.init_oracle_client(lib_dir=os.path.join(os.getenv("HOME"), "Downloads", "instantclient_19_8"))
 
# import json_data
from sqlalchemy import (
        Column,
        Identity,
        Integer,
        String,
        create_engine,
)
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
 
Base = declarative_base()
 
class Bleh(Base):
    __tablename__ = "bleh"
    id = Column(Integer, Identity(on_null=True), primary_key=True)
    othercol = Column(String(255))
 
if __name__ == "__main__":
    host = 'localhost'
    port = 1521
    user = '<bleh>'
    password = '<bleh>'
    sid = cx_Oracle.makedsn(host, port, service_name='<mycdb>')
    sql_engine = create_engine(f'oracle://{user}:{password}@{sid}',
                                    convert_unicode=False,
                                    pool_recycle=10,
                                    pool_size=50,
                                    echo=True,
                                    implicit_returning=False,
                                    future=True)
 
    Base.metadata.drop_all(sql_engine)
    Base.metadata.create_all(sql_engine)
 
    Session = sessionmaker(bind=sql_engine)
 
    session = Session()
 
    instance = Bleh()
    instance.othercol = "hi there"
 
    session.add(instance)
    session.commit()
```

### Error

```
(testvenv) $ ./FlanToDbLoader/mvp.py                                
2021-08-18 17:19:15,826 INFO sqlalchemy.engine.Engine select sys_context( 'userenv', 'current_schema' ) from dual
2021-08-18 17:19:15,827 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-08-18 17:19:15,836 INFO sqlalchemy.engine.Engine select value from nls_session_parameters where parameter = 'NLS_NUMERIC_CHARACTERS'
2021-08-18 17:19:15,836 INFO sqlalchemy.engine.Engine [raw sql] {}
2021-08-18 17:19:15,839 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-08-18 17:19:15,840 INFO sqlalchemy.engine.Engine SELECT table_name FROM all_tables WHERE table_name = :name AND owner = :schema_name
2021-08-18 17:19:15,840 INFO sqlalchemy.engine.Engine [generated in 0.00014s] {'name': 'BLEH', 'schema_name': '<redacted>'}
2021-08-18 17:19:15,872 INFO sqlalchemy.engine.Engine 
DROP TABLE bleh
2021-08-18 17:19:15,872 INFO sqlalchemy.engine.Engine [no key 0.00024s] {}
2021-08-18 17:19:15,898 INFO sqlalchemy.engine.Engine COMMIT
2021-08-18 17:19:15,901 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-08-18 17:19:15,902 INFO sqlalchemy.engine.Engine SELECT table_name FROM all_tables WHERE table_name = :name AND owner = :schema_name
2021-08-18 17:19:15,902 INFO sqlalchemy.engine.Engine [cached since 0.06229s ago] {'name': 'BLEH', 'schema_name': 'PS_USER'}
2021-08-18 17:19:15,904 INFO sqlalchemy.engine.Engine 
CREATE TABLE bleh (
	id INTEGER GENERATED BY DEFAULT ON NULL AS IDENTITY, 
	othercol VARCHAR2(255 CHAR), 
	PRIMARY KEY (id)
)


2021-08-18 17:19:15,904 INFO sqlalchemy.engine.Engine [no key 0.00012s] {}
2021-08-18 17:19:15,918 INFO sqlalchemy.engine.Engine COMMIT
2021-08-18 17:19:15,923 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-08-18 17:19:15,925 INFO sqlalchemy.engine.Engine INSERT INTO bleh (othercol) VALUES (:othercol)
2021-08-18 17:19:15,926 INFO sqlalchemy.engine.Engine [generated in 0.00026s] {'othercol': 'hi there'}
2021-08-18 17:19:15,935 INFO sqlalchemy.engine.Engine ROLLBACK
Traceback (most recent call last):
  File "/Users/dmkoenig/dev/db-table-builder/FlanToDbLoader/mvp.py", line 50, in <module>
    session.commit()
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1428, in commit
    self._transaction.commit(_to_root=self.future)
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 829, in commit
    self._prepare_impl()
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 808, in _prepare_impl
    self.session.flush()
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3298, in flush
    self._flush(objects)
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3438, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3404, in _flush
    flush_context.finalize_flush_changes()
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 477, in finalize_flush_changes
    self.session._register_persistent(other)
  File "/Users/dmkoenig/.pyenv/versions/testvenv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 2465, in _register_persistent
    raise exc.FlushError(
sqlalchemy.orm.exc.FlushError: Instance <Bleh at 0x1262b0e80> has a NULL identity key.  If this is an auto-generated value, check that the database table allows generation of new primary key values, and that the mapped Column object is configured to expect these generated values.  Ensure also that this flush() is not occurring at an inappropriate time, such as within a load() event.
```

This fails with `Identity()` and `Identity(on_null=True)` with the same output. It succeeds with SQLite, and it succeeds if you replace `Identity()` with `Sequence('idseq')`.

### Versions

- OS: macOS 11.5.2
- Python: 3.9.6
- SQLAlchemy: sqlalchemy==1.4.22
- Database: Oracle 12.1.0.2, cx_Oracle==8.2.1
- DBAPI: No relevant packages installed in my site-packages. Whatever you get in a fresh 3.9.6 virtualenv

**Comment by zzzeek** [comment, marked as answer]
this is because you are using an unusual setting called "implicit_returning=False".  SQLAlhcemy has no way to get the newly generated value of an IDENTITY key if you turn off RETURNING.   remove the use of this flag to resolve the issue.

**Comment by koenigdavidmj** [comment]
That was the problem.

Do you think this is worth some additional documentation? The existing docs tell me what the effect is on the SQL, but not why I might want that. I could see a couple different paths:

* Just a warning saying "you probably don't want this in new code"
* "If you do this, then (slight benefit), but here's an incomplete list of things that might break"

**Comment by CaselIT** [comment]
I do think it's a bug, since with `"implicit_returning=False"` the session should manually fetch the generated value like it does in the db that don't support returning.

Thoughts @zzzeek?

On the other hand, why do we even have the `implicit_returning` flag? When one would want to turn that off?

**Comment by zzzeek** [comment]
> I do think it's a bug, since with `"implicit_returning=False"` the session should manually fetch the generated value like it does in the db that don't support returning.

SQLAlchemy only has three methods of fetching IDs, there is "run the default ahead of time", "use returning", and "use cursor.lastrowid".    The "run the default ahead of time" technique was always a little awkward for PKs and it is now virtually never used, i'd have to check what its role is for non-PK defaults but I would favor taking that whole thing out.

> 
> 
> On the other hand, why do we even have the `implicit_returning` flag? When one would want to turn that off?

there's really no reason to ever turn this flag off, it is a remnant of when DBs were first beginning to support RETURNING and SQLAlchemy's support was very new.    I'd want to do a search to see if there are perhaps some MSSQL cases that people still want it (these would be driver-level issues), but for Oracle, the cx_Oracle driver is now a very solid, super-well supported driver and we are actually behind them on our support for RETURNING so I dont see any reason this flag is needed; we have some docs about this flag in the oracle docs so it's likely that when we were on cx_oracle 4 and 5 and it was hard to get support / docs for the driver, there were probably problems.

in short, there's no need for this flag im aware of so it should just be removed w/ a deprecation phase.

**Comment by CaselIT** [comment]
Since it's a pk in the example above, shouldn't this case fall back on "use cursor.lastrowid"? Not sure if maybe something is messed up in the pk detection when we use identity

**Comment by CaselIT** [comment]
> 
> 
> Since it's a pk in the example above, shouldn't this case fall back on "use cursor.lastrowid"? Not sure if maybe something is messed up in the pk detection when we use identity

I mean we currently don't have a db that does support identity but does not support returning (oracle, pg and mssql all have returning), so maybe we have a bug there

**Comment by zzzeek** [comment]
cursor.lastrowid is not a thing for PG and Oracle, and only marginally for MSSQL.    the code relies on RETURNING for these DBs and im fine moving towards that being the only way we operate on these DBs for now.

**Comment by CaselIT** [comment]
Ok. I guess we can move to discussion then
